### PR TITLE
Gracefully skip reference indexing for objects with a null ParentalKey

### LINKED ISSUE #9583: Wagtail 4.1: modelcluster.fields.ParentalKey throws AttributeError: 'NoneType' object has no attribute '_meta'
I'm been trying to update my blog to Wagtail 4.1 and I'm having issues with the following models.py setup:

```python
from modelcluster.models import ClusterableModel


@register_snippet
class BlogCategory(ClusterableModel):
    """
    A post category. Every blog post can only be in one category
    """

    name = models.CharField(max_length=255)

    panels = [ FieldPanel("name") ]

    class Meta:
        verbose_name = _("Category")
        verbose_name_plural = _("Categories")


class BlogPage(BasePage):
    """
    A Blog Page
    """
   # Other fields here

    category = ParentalKey(
        "BlogCategory",
        related_name="categories",
        null=True,
        blank=True,
        on_delete=models.SET_NULL,
    )
```

This code, for example, fails:

```
blog = BlogIndexPage(title="Blog", slug="blog")
post = BlogPage(title="post")
blog.add_child(instance=post)
```

The source of the error is[ this change in the `signal_handlers.py` module](https://github.com/wagtail/wagtail/commit/212b2e19a680410ac1ab3fe4f18cdcb318afaa67#diff-9e2fc90f09e08c4af86e25111803a306e91d17f5ad97333b51a2f2805226af2a) that was added on Wagtail 4.1.

More specifically, this signal:

```python
def update_reference_index_on_save(instance, **kwargs):
    # Don't populate reference index while loading fixtures as referenced objects may not be populated yet
    if kwargs.get("raw", False):
        return

    if getattr(reference_index_auto_update_disabled, "value", False):
        return

    # If the model is a child model, find the parent instance and index that instead
    while True:
        parental_keys = list(
            filter(
                lambda field: isinstance(field, ParentalKey),
                instance._meta.get_fields(),
            )
        )
        if not parental_keys:
            break

        instance = getattr(instance, parental_keys[0].name)

    if ReferenceIndex.model_is_indexable(type(instance)):
        with transaction.atomic():
            ReferenceIndex.create_or_update_for_object(instance)
```

Since an instance with the `parental_keys` var is populated even when the field is `null`, the `instance` variable becomes null and we get an error. 

### Technical details

-   Python version: Python 3.10.8
-   Django version: Django 4.1.6
-   Wagtail version: Wagtail 4.1

### PR DESCRIPTION
**Pull request by gasman**
Fixes #9583. The use case for null ParentalKeys is a bit questionable, but if they exist, we don't want reference indexing to break on them. Since references are always stored against the parent object, there's no valid way to record these references, so gracefully skip over them instead.

### CODE REVIEW

File: wagtail/signal_handlers.py [hunk 1]
```diff
@@ -83,6 +83,9 @@ def update_reference_index_on_save(instance, **kwargs):
             break
 
         instance = getattr(instance, parental_keys[0].name)
+        if instance is None:
```
Comment (allcaps) [inline wagtail/signal_handlers.py:86]: Should we rename `instance` to `parent`? This name might be confusing.

Or am I missing something?

**Review by allcaps — CHANGES_REQUESTED** [review]
Just a nitpick about the variable name.

File: wagtail/signal_handlers.py (hunk 1 above)
Comment (gasman) [reply wagtail/signal_handlers.py:86]: This code is inside a loop where the parent from the current iteration becomes the `instance` for the next.

We could do something like:
```
        parent = getattr(instance, parental_keys[0].name)
        if parent is None:
            # there is no valid object to record references against
            return
        instance = parent
```

but I think the comment already provides the necessary clarification.

**Comment by gasman** [comment]
Merged via #9625

**Pull request closed without merging**